TST: add normalize option to assert methods + fix tests for GEOS 3.9

jorisvandenbossche opened the pull request:
Closes #1818

This adds `normalize` capabilities. For now I only added it to our assert_.. functions. Eventually we might also want to add it to the GeometryArray/GeoSeries, but for that I would like to wait on Shapely including it (https://github.com/Toblerity/Shapely/pull/1090/)

Normalizing the geometries fixes most of the overlay tests, only the NYBB tests are still failing (here the differences with GEOS 3.9 are bigger than just some coordinate order). Not sure what the best way to solve this is. We can skip the geometry equality check for now with GEOS 3.9 (we already check area and bounds, and that passes), and later re-produce the expected files with GEOS 3.9

martinfleis commented:
Thanks a lot!

> Eventually we might also want to add it to the GeometryArray/GeoSeries, but for that I would like to wait on Shapely including it (Toblerity/Shapely#1090)

Agree.

> We can skip the geometry equality check for now with GEOS 3.9 (we already check area and bounds, and that passes), and later re-produce the expected files with GEOS 3.9

Let's do that. I have also observed some differences in how GEOS 3.9 handles specific situations (e.g. order of geometries from polygonize is different) and concluded that it is the best solution to tailor tests for that.

jorisvandenbossche commented:
> I have also observed some differences in how GEOS 3.9 handles specific situations (e.g. order of geometries from polygonize is different)

If it's only the order of geometries, that's something that `normalize` should solve. But in this overlay case with NYBB, also the number of geometries of a MultiPolygon is different, which is not something normalize fixes.

martinfleis commented:
It is not an order of geometries within MultiPolygon, it is an order of Polygons from `polygonize`.

> the number of geometries of a MultiPolygon is different

That is weird. Have you looked at the difference in detail?

martinfleis commented:
It seems that our `sort` keyword in STRTree is now obsolete since the result is sorted in any case with GEOS 3.9 - https://github.com/geopandas/geopandas/pull/1826/checks?check_run_id=1901961534#step:7:2912

jorisvandenbossche commented:
> > the number of geometries of a MultiPolygon is different
>
> That is weird. Have you looked at the difference in detail?

That's actually already the case before GEOS 3.9 as well (that's the reason we *already* were skipping many polygons when asserting the actual geometries in the nybb test)

Looking at the actual vs expected result on GEOS 3.9 in a notebook, it are all small but real differences at the border of the circles. 

New result:

```
result.plot(figsize=(10, 10))
```

![image](https://user-images.githubusercontent.com/1020496/108121485-2b197180-70a3-11eb-9c41-bd23b2e76a8c.png)

Difference with result/expected:

```
diffs = result.difference(expected)
diffs[~diffs.is_empty].plot(edgecolor='k', figsize=(10, 10))
```

![image](https://user-images.githubusercontent.com/1020496/108121580-4dab8a80-70a3-11eb-9a63-9ead4227b760.png)

Seeing this image, with mostly small change at the border (even in areas that are otherwise fully contained in a polygon of the boroughs), I am wondering: could this be caused because the *buffer* is creating a slightly different circular polygon? Because the expected results were generated with buffers created with an older GEOS version, so if with the latest GEOS we get slightly different buffers, it's normal we can't expect the same result now. 

I don't know why the output of buffer would have changed, but so that's actually the same issue as with the `test_polygon_patch` where the number of vertices of the buffer changed (now I think of it)

jorisvandenbossche commented:
(I should test my hypothesis by creating and saving the buffers to a file with an older GEOS version, and reading those in instead of creating on the fly in the test)

martinfleis commented:
Good call with buffers, you're most likely right that it creates them differently now. Let's skip for now and change these tests later to be a bit more forgiving.

jorisvandenbossche commented:
Turns out we already have the buffers included as a shapefile ;) So easy enough to use those instead, and that also solves the additional differences with GEOS 3.9.

That only leaves the pre-existing differences with the QGIS output. I looked at the difference for "union", and it's a really tiny difference:

![image](https://user-images.githubusercontent.com/1020496/108173450-6eed9480-70fe-11eb-9088-5f3845527aaa.png)

![image](https://user-images.githubusercontent.com/1020496/108173502-7ad95680-70fe-11eb-8e82-0d457a14e5f1.png)

So the right figures show the result from geopandas, and the expected QGIS result doesn't have this one sliver that sticks out.

jorisvandenbossche commented:
> It seems that our sort keyword in STRTree is now obsolete since the result is sorted in any case with GEOS 3.9 

I don't think it's actually sorted, but just a different order (our tests currently relied on testing the order we got we previous GEOS versions for `sort=False`, and that now changed apparently):

```
>           assert_array_equal(res, expected)
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatched elements: 2 / 3 (66.7%)
E           Max absolute difference: 1
E           Max relative difference: 1.
E            x: array([1, 0, 2])
E            y: array([0, 1, 2])
```

(it's the expected that was "sorted", and the result is no longer)

Updated the tests to first check the sorted elements are the same, and then let the exact equality also fail for pygeos.

jorisvandenbossche commented:
Aha, all green now! (for some reason this morning the dev build hang for a couple of hours...)
